pkg/controllers/provisioning/scheduling/scheduler.go

```diff
@@ -50,60 +50,77 @@ func init() {
 }
 
 type Scheduler struct {
-	kubeClient client.Client
-	topology   *Topology
+	kubeClient  client.Client
+	preferences *Preferences
 }
 
 func NewScheduler(kubeClient client.Client) *Scheduler {
 	return &Scheduler{
-		kubeClient: kubeClient,
-		topology:   &Topology{kubeClient: kubeClient},
+		kubeClient:  kubeClient,
+		preferences: NewPreferences(),
 	}
 }
 
-func (s *Scheduler) Solve(ctx context.Context, provisioner *v1alpha5.Provisioner, instanceTypes []cloudprovider.InstanceType, pods []*v1.Pod) ([]*Node, error) {
+func (s *Scheduler) Solve(ctx context.Context, provisioner *v1alpha5.Provisioner, instanceTypes []cloudprovider.InstanceType, pods []*v1.Pod) (nodes []*Node, err error) {
 	defer metrics.Measure(schedulingDuration.WithLabelValues(injection.GetNamespacedName(ctx).Name))()
 	constraints := provisioner.Spec.Constraints.DeepCopy()
 
 	sort.Slice(pods, byCPUAndMemoryDescending(pods))
 	sort.Slice(instanceTypes, byPrice(instanceTypes))
 
-	// Inject temporarily adds specific NodeSelectors to pods, which are then
-	// used by scheduling logic. This isn't strictly necessary, but is a useful
-	// trick to avoid passing topology decisions through the scheduling code. It
-	// lets us treat TopologySpreadConstraints as just-in-time NodeSelectors.
-	if err := s.topology.Inject(ctx, constraints, pods); err != nil {
-		return nil, fmt.Errorf("injecting topology, %w", err)
-	}
+	topology := NewTopology(s.kubeClient, constraints)
 
-	nodeSet, err := NewNodeSet(ctx, constraints, s.kubeClient)
+	nodeSet, err := NewNodeSet(ctx, topology, constraints, instanceTypes, s.kubeClient)
 	if err != nil {
 		return nil, fmt.Errorf("constructing nodeset, %w", err)
 	}
 
-	unschedulableCount := 0
-	for _, pod := range pods {
-		isScheduled := false
-		for _, node := range nodeSet.nodes {
-			if err := node.Add(pod); err == nil {
-				isScheduled = true
-				break
+	if err := topology.TrackTopologies(ctx, pods...); err != nil {
+		return nil, fmt.Errorf("tracking topology counts, %w", err)
+	}
+
+	lastErrors := map[*v1.Pod]error{}
```

Comment:
Not obvious what `last` means until you read the impl below. I'd probably lean the former, since we're already in the scheduling loop so it can be inferred from context.
```suggestion
	errs := map[*v1.Pod]error{}
```
```suggestion
	schedulingErrs := map[*v1.Pod]error{}
```